README
LiblineaR is a wrapper around the LIBLINEAR C/C++ library for machine learning.
LIBLINEAR is a simple library for solving large-scale regularized linear 
classification and regression. It currently supports 
- L2-regularized logistic regression/L2-loss support vector classification (SVM)/
    L1-loss support vector classification (SVM)
- L1-regularized L2-loss support vector classification/L1-regularized logistic regression
- L2-regularized L2-loss support vector regression/L1-loss support vector regression. 

The main features of LiblineaR include multi-class classification (one-vs-the rest, and
Crammer & Singer method), cross validation for model selection, probability
estimates (logistic regression only) or weights for unbalanced data. The
estimation of the models is particularly fast as compared to other libraries.

The two first blocks of the package version indicates which version of LIBLINEAR is
currently supported by LiblineaR. For example: 1.32-14 means that the package supports
the version 1.32 of LIBLINEAR.

For more information on the C/C++ LIBLINEAR library itself, refer to:

R.-E. Fan, K.-W. Chang, C.-J. Hsieh, X.-R. Wang, and C.-J. Lin. 
LIBLINEAR: A Library for Large Linear Classification, 
Journal of Machine Learning Research 9(2008), 1871-1874. 

Software available at http://www.csie.ntu.edu.tw/~cjlin/liblinear .

Copyright
---------

All of this software is copyrighted by the list of authors in the DESCRIPTION file of 
the package and subject to the GNU GENERAL PUBLIC LICENSE, Version 2, see the file
COPYING for details. The LIBLINEAR C/C++ code is copyright Chih-Chung Chang and 
Chih-Jen Lin.

Installation
------------

This package should be installed using the `R CMD INSTALL' mechanism, see
the R online help or R manuals for details on how to install packages.

Mathematical Details
--------------------

Formulations:

For L2-regularized logistic regression (-s 0), we solve

min_w w^Tw/2 + C \sum log(1 + exp(-y_i w^Tx_i))

For L2-regularized L2-loss SVC dual (-s 1), we solve

min_alpha  0.5(alpha^T (Q + I/2/C) alpha) - e^T alpha
    s.t.   0 <= alpha_i,

For L2-regularized L2-loss SVC (-s 2), we solve

min_w w^Tw/2 + C \sum max(0, 1- y_i w^Tx_i)^2

For L2-regularized L1-loss SVC dual (-s 3), we solve

min_alpha  0.5(alpha^T Q alpha) - e^T alpha
    s.t.   0 <= alpha_i <= C,

For L1-regularized L2-loss SVC (-s 5), we solve

min_w \sum |w_j| + C \sum max(0, 1- y_i w^Tx_i)^2

For L1-regularized logistic regression (-s 6), we solve

min_w \sum |w_j| + C \sum log(1 + exp(-y_i w^Tx_i))

For L2-regularized logistic regression (-s 7), we solve

min_alpha  0.5(alpha^T Q alpha) + \sum alpha_i*log(alpha_i) + \sum (C-alpha_i)*log(C-alpha_i) - a constant
    s.t.   0 <= alpha_i <= C,

where

Q is a matrix with Q_ij = y_i y_j x_i^T x_j.

For L2-regularized L2-loss SVR (-s 11), we solve

min_w w^Tw/2 + C \sum max(0, |y_i-w^Tx_i|-epsilon)^2

For L2-regularized L2-loss SVR dual (-s 12), we solve

min_beta  0.5(beta^T (Q + lambda I/2/C) beta) - y^T beta + epsilon \sum |beta_i|

For L2-regularized L1-loss SVR dual (-s 13), we solve

min_beta  0.5(beta^T Q beta) - y^T beta + epsilon \sum |beta_i|
    s.t.   -C <= beta_i <= C,

where

Q is a matrix with Q_ij = x_i^T x_j.

If bias >= 0, w becomes [w; w_{n+1}] and x becomes [x; bias].

The primal-dual relationship implies that -s 1 and -s 2 give the same
model, -s 0 and -s 7 give the same, and -s 11 and -s 12 give the same.

We implement 1-vs-the rest multi-class strategy for classification. 
In training i vs. non_i, their C parameters are (weight from -wi)*C 
and C, respectively. If there are only two classes, we train only one
model. Thus weight1*C vs. weight2*C is used. See examples below.

We also implement multi-class SVM by Crammer and Singer (-s 4):

min_{w_m, \xi_i}  0.5 \sum_m ||w_m||^2 + C \sum_i \xi_i
    s.t.  w^T_{y_i} x_i - w^T_m x_i >= \e^m_i - \xi_i \forall m,i

where e^m_i = 0 if y_i  = m,
      e^m_i = 1 if y_i != m,

Here we solve the dual problem:

min_{\alpha}  0.5 \sum_m ||w_m(\alpha)||^2 + \sum_i \sum_m e^m_i alpha^m_i
    s.t.  \alpha^m_i <= C^m_i \forall m,i , \sum_m \alpha^m_i=0 \forall i

where w_m(\alpha) = \sum_i \alpha^m_i x_i,
and C^m_i = C if m  = y_i,
    C^m_i = 0 if m != y_i.


NEWS
Note: In this file, LiblineaR refers to this R package, while LIBLINEAR refers to the original C/C++ library.

1.94-2: 2015/01/30
	1. Cleaned DESCRIPTION

	2. library(SparseM) replaced by 'require' or 'requireNamespace' as appropriate.

1.94-1: 2015/01/30
	1. Upgraded to LIBLINEAR version 1.94, including support vector *regression*
		- argument labels of function LiblineaR was renamed as target. The old naming is still accepted with a warning
		- argument type of function LiblineaR may take extra values (11,12,13)
		- extra argument svr_eps of function LiblineaR for the tolerance of regression loss
		- if class labels are -1 and 1, ensure that positive decision values represent the class 1
		- enriched examples, with cases of regression

	2. Fix CITATION file format to satisfy CRAN requirements

	3. Cleaning and reformatting of man pages

	4. Fixing a bug in class labels ordering 
	
1.80-11: 2014/09/18
	1. Refactored C code in order to facilitate upgrades of LIBLINEAR.
	
	2. Fix memory leak in predictLinear (the buffer x wasn't freed)
	
1.80-10: 2014/09/14
	1. Shortened examples runtime.
	
	2. Uniformized notations in DESCRIPTION file for LiblineaR (this package) and LIBLINEAR (the C/C++ library wrapped by this package). 

1.80-9: 2014/09/12
	1. Replaced all expressions: 

		rand()%(a-b); 
	by 

		GetRNGstate();
		(int) (unif_rand()*(a - b))%(a - b);
		PutRNGstate();

	in linear.cpp, where a and b might take different values.

1.80-8: 2014/09/11

	1. Corrected a bug in memory allocation for sparse matrices:
		src/trainLinear.c

		These changes are to be credited to Christian Wolf.

	2. Added a PACKAGE argument to speed up call to .C(...):
		R/LiblineaR.R
		R/predict.R

1.80-7: 2013/06/24

	1. Modification of the following files in order to support sparse matrices:
		R/LiblineaR.R
		R/predict.R
		src/predictLinear.c 
		src/trainLinear.c

		These changes are mainly to be credited to Kai-Hsiang Hsu, from the Department of Computer Science of the National Taiwan University.

	2. Addition of examples to reflect the use of sparse matrices in the following file:
		man/LiblineaR.Rd

1.80-6: 2013/03/26
	
	1. Corrected a memory mapping bug in predictLinear.c

1.80-5: 2013/03/25

	1. Suppress printing to stdout in linear.cpp, tron.cpp
	
	2. Suppress the use of exit(1); in predictLinear.c

	3. Correct a bug when retrieving weights from C to R for multi-class models
 
1.80-4: 2011/04/21

	1. Correct bugs in linear.cpp (update of solve_l1r_l2_svc and solve_l1r_lr)

1.80-3: 2011/04/20
	
	1. Add:

	extern "C" 

	before each function below :

	//
	// Interface function
	//
	
	in linear.cpp

	2. Suppress useless functions:

	- save_model
	- load_model
	
	in:
	
	linear.cpp
	linear.h	

	3. change all "fprintf" into Rprintf in :
	linear.cpp
	trainLinear.c
	predictLinear.c

	4. change:

	#define Malloc(type,n) (type *)malloc((n)*sizeof(type)) 

	into 

	#define Malloc(type,n) (type *)Calloc(n,type)

	in:
	trainLinear.c
	linear.cpp

	5. replace all the malloc((n)*sizeof(type)) by Calloc(n,type) in predictLinear.c

	5. replace all the realloc(*p,size) by Realloc(*p,n,type) in linear.cpp

	7. replace all free() by Free() in:

	trainLinear.c
	predictLinear.c
	linear.cpp

	8. Add
	
	#include <R.h>
	#include <Rinternals.h>
	#include <R_ext/Rdynload.h>
	
	in linear.cpp

1.80-2: 2011/04/12

	1. Incorporate changes from LIBLINEAR versions 1.51 to 1.80:
		- Use set_print_string_function to set the print function
		- Add free_model_content and free_and_destroy_model functions (avoid memory problem if users declare a model variable)
		- Add check_probability_model (consistent with libsvm)
		- A new solver: coordinate descent for dual logistic regression
		- New optimization method for l1-regularized logistic regression
		- linear.cpp: 
			* Use 1-norm stopping condition for l1-regularized solvers 
			* newton_iter < l/10 replaced by newton_iter <= l/10 in l2r_lr_dual (for l < 10)
		
	2. predict.LiblineaR function can return additional information:
		- probabilities (only for logistic regression models)
		- decision values




R/LiblineaR.ACF.R
### Documentation ####
#' Linear predictive models estimation with Online Adaptation of Coordinate 
#' Frequencies based on the LIBLINEAR C/C++ Library.
#' 
#' @description
#' \code{LiblineaR.ACF} is a modification of the LiblineaR package that
#' uses the idea of adaptive coordinate frequencies (ACF) method.
#' Solving the linear SVM problem with coordinate descent
#' is very efficient and is implemented in one of the most often used packages,
#' LIBLINEAR (available at http://www.csie.ntu.edu.tw/~cjlin/liblinear). 
#' It has been shown that the uniform selection of coordinates can be 
#' accelerated by using an online adaptation of coordinate frequencies (ACF).
#' This package implements ACF and is based on LIBLINEAR as well as
#' the LiblineaR package (https://cran.r-project.org/package=LiblineaR).
#' It currently supports L2-regularized L1-loss as well as L2-loss linear SVM. 
#' Similar to LIBLINEAR multi-class classification (one-vs-the rest, and 
#' Crammer & Singer method) and cross validation for model selection is 
#' supported. The training of the models based on ACF is much faster than 
#' standard LIBLINEAR on many problems.
#' 
#' @details
#' For details for the implementation of LIBLINEAR, see the README file of the
#' original c/c++ LIBLINEAR library at
#' \url{http://www.csie.ntu.edu.tw/~cjlin/liblinear}.
#' The ACF code can be found 
#' 
#' @param data 	a nxp data matrix. Each row stands for an example (sample,
#'   point) and each column stands for a dimension (feature, variable). A sparse
#'   matrix (from SparseM package) will also work.
#' @param target a response vector for prediction tasks with one value for 
#'   each of the n rows of \code{data}. For classification, the values 
#'   correspond to class labels and can be a 1xn matrix, a simple vector or a 
#'   factor. For regression, the values correspond to the values to predict, and
#'   can be a 1xn matrix or a simple vector.
#' @param type \code{LiblineaR} can produce 10 types of (generalized) linear 
#'   models, by combining several types of loss functions and regularization 
#'   schemes. The regularization can be L1 or L2, and the losses can be the
#'   regular L2-loss for SVM (hinge loss), L1-loss for SVM, or the logistic loss
#'   for logistic regression. The default value for \code{type} is 0. See
#'   details below. Valid options are:
#'   \describe{
#'     \item{for multi-class classification}{
#'       \itemize{
#'         \item 1 -- L2-regularized L2-loss support vector classification (dual)
#'         \item 3 -- L2-regularized L1-loss support vector classification (dual)
#'         \item 4 -- support vector classification by Crammer and Singer
#'       }
#'     }
#'   }
#' 
#' @param cost cost of constraints violation (default: 1). Rules the trade-off
#'   between regularization and correct classification on \code{data}. It can be
#'   seen as the inverse of a regularization constant. See information on the
#'   'C' constant in details below. A usually good baseline heuristics to tune
#'   this constant is provided by the \code{heuristicC} function in the LiblineaR
#'   package.
#' @param epsilon set tolerance of termination criterion for optimization.
#'   If \code{NULL}, the LIBLINEAR defaults are used, which are:
#'   \describe{
#'     \item{if \code{type} is 1, 3 or 4}{\code{epsilon}=0.1}
#'   }
#'   
#'   % Watch out, below: in fact, \eqn{LaTeX}{ascii} is counter-intuitive.
#'   % the man pages in terminal mode are extracted from LaTeX argument, not the ascii argument !
#'   
#'   The meaning of \code{epsilon} is as follows:
#'   \describe{
#' 	    \item{if \code{type} is 0 or 2:}{
#' 			      \eqn{|f'(w)|_{2} \le \code{epsilon} \times \min (pos,neg) / l \times |f'(w_{0})|_{2}}{|f'(w)|_2 \le  epsilon * min(pos,neg) / l *|f'(w0)|_2},
#' 			      where f is the primal function and pos/neg are # of positive/negative data (default 0.01)}
#' 	    \item{if \code{type} is 11:}{
#' 			      \eqn{|f'(w)|_{2} \le \code{epsilon} \times |f'(w_{0})|_{2},}{|f'(w)|_2 \le epsilon * |f'(w0)|_2,}
#' 			      where f is the primal function (default 0.001)}
#'      \item{if \code{type} is 1, 3, 4 or 7:}{
#'            Dual maximal violation \eqn{\le \code{epsilon}}{\le epsilon} 
#'            (default 0.1)}
#'      \item{if \code{type} is 5 or 6:}{
#' 			      \eqn{|f'(w)|_\infty \le \code{epsilon}\times \min(pos,neg)/l\ |f'(w_{0})|_\infty,}{|f'(w)|_inf \le epsilon * min(pos,neg) / l*|f'(w0)|_inf,}
#' 			      where f is the primal function (default 0.01)}
#'      \item{if \code{type} is 12 or 13:}{
#' 			      \eqn{|f'(\alpha)|_1 \le \code{epsilon}\times |f'(\alpha_{0})|_1,}{|f'(alpha)|_1 \le epsilon * |f'(alpha0)|_1,}
#' 			      where f is the dual function (default 0.1)}
#' 	}
#' @param svr_eps set tolerance margin (epsilon) in regression loss function of SVR. Not used for classification methods.
#' @param bias if \code{bias} is \code{TRUE} (default), instances of \code{data} becomes [\code{data}; 1].
#' @param wi a named vector of weights for the different classes, used for
#'   asymmetric class sizes. Not all factor levels have to be supplied (default
#'   weight: 1). All components have to be named according to the corresponding
#'   class label. Not used in regression mode.
#' @param cross if an integer value k>0 is specified, a k-fold cross validation
#'   on \code{data} is performed to assess the quality of the model via a
#'   measure of the accuracy. Note that this metric might not be appropriate if
#'   classes are largely unbalanced. Default is 0.
#' @param change_rate learning rate of the preference adaptation, default is 0.2
#' @param pref_min lower bound on the preference adaptation, default is 1/20
#' @param pref_max upper bound on the preference adaptation, default is 20
#' @param max_iter the maximum number of iterations, default (from original LIBLINEAR code) is 1000.
#' @param verbose if \code{TRUE}, information are printed. Default is
#'   \code{FALSE}.
#' @param ... for backwards compatibility, parameter \code{labels} may be
#'   provided instead of \code{target}. A warning will then be issued, or an
#'   error if both are present. Other extra parameters are ignored.
#' 
#' @return
#' 	If \code{cross}>0, the average accuracy (classification) or mean square error (regression) computed over \code{cross} runs of cross-validation is returned.\cr\cr
#' Otherwise, an object of class \code{"LiblineaR"} containing the fitted model is returned, including:
#' \item{TypeDetail}{A string decsribing the type of model fitted, as determined by \code{type}.}
#' \item{Type}{An integer corresponding to \code{type}.}
#' \item{W}{A matrix with the model weights. If \code{bias} is TRUE, \code{W} contains p+1 columns, the last being the bias term. The columns are named according to the names of \code{data}, if provided, or \code{"Wx"} where \code{"x"} ranges from 1 to the number of dimensions. The bias term is named \code{"Bias"}.If the number of classes is 2, or if in regression mode rather than classification, the matrix only has one row. If the number of classes is k>2 (classification), it has k rows. Each row i corresponds then to a linear model discriminating between class i and all the other classes. If there are more than 2 classes, rows are named according to the class i which is opposed to the other classes.}
#' \item{Bias}{TRUE or FALSE, according to the value of \code{bias}}
#' \item{ClassNames}{A vector containing the class names. This entry is not returned in case of regression models.}
#' 
#' @references
#' 	\itemize{
#' \item 
#' For more information on LIBLINEAR itself, refer to:\cr
#' R.-E. Fan, K.-W. Chang, C.-J. Hsieh, X.-R. Wang, and C.-J. Lin.\cr
#' \emph{LIBLINEAR: A Library for Large Linear Classification,}\cr
#' Journal of Machine Learning Research 9(2008), 1871-1874.\cr
#' \url{http://www.csie.ntu.edu.tw/~cjlin/liblinear}
#' }
#' 
#' @author Aydin Demircioglu \email{aydin.demircioglu@@ini.rub.de}
#'   Based on LiblineaR package by Thibault Helleputte \email{thibault.helleputte@@dnalytics.com} and\cr
#'   Pierre Gramme \email{pierre.gramme@@dnalytics.com}.\cr 
#'   Based on C/C++-code by Chih-Chung Chang and Chih-Jen Lin
#'   Based on C/C++-code by Tobias Glasmachers and Urun Dogan
#' 
#' @note Classification models usually perform better if each dimension of the data is first centered and scaled.
#' 
#' @seealso \code{\link{predict.LiblineaR.ACF}}, \code{\link[LiblineaR]{heuristicC}}
#' 
#' @examples
#' data(iris)
#' attach(iris)
#' 
#' x=iris[,1:4]
#' y=factor(iris[,5])
#' train=sample(1:dim(iris)[1],100)
#' 
#' xTrain=x[train,]
#' xTest=x[-train,]
#' yTrain=y[train]
#' yTest=y[-train]
#' 
#' # Center and scale data
#' s=scale(xTrain,center=TRUE,scale=TRUE)
#' 
#' # Find the best model with the best cost parameter via 10-fold cross-validations
#' tryTypes=c(0:7)
#' tryCosts=c(1000,1,0.001)
#' bestCost=NA
#' bestAcc=0
#' bestType=NA
#' 
#' for(ty in tryTypes){
#' 	for(co in tryCosts){
#' 		acc=LiblineaR.ACF(data=s,target=yTrain,type=ty,cost=co,
#'				bias=TRUE,cross=5,verbose=FALSE)
#' 		cat("Results for C=",co," : ",acc," accuracy.\n",sep="")
#' 		if(acc>bestAcc){
#' 			bestCost=co
#' 			bestAcc=acc
#' 			bestType=ty
#' 		}
#' 	}
#' }
#' 
#' cat("Best model type is:",bestType,"\n")
#' cat("Best cost is:",bestCost,"\n")
#' cat("Best accuracy is:",bestAcc,"\n")
#' 
#' # Re-train best model with best cost value.
#' m=LiblineaR.ACF(data=s,target=yTrain,type=bestType,cost=bestCost,bias=TRUE,verbose=FALSE)
#' 
#' # Scale the test data
#' s2=scale(xTest,attr(s,"scaled:center"),attr(s,"scaled:scale"))
#' 
#' # Make prediction
#' pr=FALSE
#' if(bestType==0 || bestType==7) pr=TRUE
#' 
#' p=predict(m,s2,proba=pr,decisionValues=TRUE)
#' 
#' # Display confusion matrix
#' res=table(p$predictions,yTest)
#' print(res)
#' 
#' # Compute Balanced Classification Rate
#' BCR=mean(c(res[1,1]/sum(res[,1]),res[2,2]/sum(res[,2]),res[3,3]/sum(res[,3])))
#' print(BCR)
#' 
#' #' #############################################
#' 
#' # Example of the use of a sparse matrix:
#' 
#' if(require(SparseM)){
#' 
#'  # Sparsifying the iris dataset:
#'  iS=apply(iris[,1:4],2,function(a){a[a<quantile(a,probs=c(0.25))]=0;return(a)})
#'  irisSparse<-as.matrix.csr(iS)
#' 
#'  # Applying a similar methodology as above:
#'  xTrain=irisSparse[train,]
#'  xTest=irisSparse[-train,]
#' 
#'  # Re-train best model with best cost value.
#'  m=LiblineaR.ACF(data=xTrain,target=yTrain,type=bestType,cost=bestCost,bias=TRUE,verbose=FALSE)
#' 
#'  # Make prediction
#'  p=predict(m,xTest,proba=pr,decisionValues=TRUE)
#' 
#'  # Display confusion matrix
#'  res=table(p$predictions,yTest)
#'  print(res)
#' }
#' 
#'
#' 
#' 
#' @keywords classif regression multivariate models optimize classes
#'
#' @export

### Implementation ####
LiblineaR.ACF<-function(data, target, type=0, cost=1, epsilon=0.01, svr_eps=NULL, bias=TRUE, wi=NULL, cross=0, 
    change_rate = 0.2, pref_min = 0.05, pref_max = 20.0, max_iter = 1000, verbose=FALSE, ...) {
	# <Arg preparation>
  
  if(sparse <- inherits(data, "matrix.csr")){
    if(requireNamespace("SparseM",quietly=TRUE)){
  		# trying to handle the sparse martix case
	  	data = SparseM::t(SparseM::t(data)) # make sure column index are sorted
		  n = data@dimension[1]
		  p = data@dimension[2]
    } else {
      stop("newx inherits from 'matrix.csr', but 'SparseM' package is not available. Cannot proceed further. Use non-sparse matrix or install SparseM.")
    }
	} else {
		# Nb samples
		n=dim(data)[1]
		# Nb features
		p=dim(data)[2]
	}

	# Backwards compatibility after renaming 'labels' to 'target'
	cc = match.call()
	if ("labels" %in% names(cc)) {
		if("target" %in% names(cc)) {
			stop("Argument 'labels' is deprecated and was renamed to 'target'. Stopping due to ambiguous call.")
		} else {
			warning("Argument 'labels' is deprecated and was renamed to 'target'.")
			target=eval(cc$labels, parent.frame())
		}
	}
	rm(cc)
	
	# Bias
	b = if(bias) 1 else -1
	
	# Type 
	typesLabels = c("L2-regularized logistic regression (primal)", "L2-regularized L2-loss support vector classification (dual)", "L2-regularized L2-loss support vector classification (primal)", "L2-regularized L1-loss support vector classification (dual)",
									"support vector classification by Crammer and Singer", "L1-regularized L2-loss support vector classification", "L1-regularized logistic regression", "L2-regularized logistic regression (dual)",
									"", "", "",
									"L2-regularized L2-loss support vector regression (primal)", "L2-regularized L2-loss support vector regression (dual)", "L2-regularized L1-loss support vector regression (dual)")
	typesLabels = gsub("[()]","",typesLabels)
	typesCodes = c("", "L2R_L2LOSS_SVC_DUAL", "", "L2R_L1LOSS_SVC_DUAL",
								 "MCSVM_CS", "", "", "",
								 "", "", "",
								 "", "", "")
	types=ifelse(typesCodes=="", "", paste0(typesLabels, " (", typesCodes, ")"))
	if(!type %in% (which(types!="")-1))
		stop("Unknown model type ",type,". Expecting one of: ", paste(which(types!="")-1, collapse=", "))
	isRegression = type>=11
	
	# Epsilon
	if(is.null(epsilon) || epsilon<0){
		# Will use LIBLINEAR default value for epsilon
		epsilon = -1
	}

	# Regression margin (epsilon)
	if(is.null(svr_eps) || svr_eps<0){
		svr_eps = 0.1
		# This default value may not make sense (depends on expected precision)
		if(isRegression)
			warning("No value provided for svr_eps. Using default of 0.1")
	}
	
	# Target
	if(!is.null(dim(target))) {
		if(identical(dim(target), c(n,1)))
			target = target[,1]
		else
			stop("Wrong dimension for target")
	}
	if(length(target)!=n){
		stop("Number of target elements disagrees with number of data instances.")
	}
	
	if(isRegression) {
		yC = as.double(target)
		isMulticlass = FALSE
		targetLabels = NULL
		nbClass = 2 # necessary for predict
		nrWi = 0
		Wi = NULL
		WiLabels = NULL
	} else {
		if(is.character(target))
			target = factor(target)
		
		# targetLabels are sorted by first occurrence in target ; if target is a factor, targetLabels too, with the same levels.
		targetLabels = unique(target)
		nbClass = length(targetLabels)
		
		yC = as.integer(target)
		
		if (nbClass<2)
			stop("Wrong number of classes ( < 2 ).")
		isMulticlass = (nbClass>2) || (nbClass==2 && type==4)
		
		# Different class penalties? Default to 1
		nrWi=nbClass
		Wi=rep(1,times=nrWi)
		names(Wi)=as.character(targetLabels)
		WiLabels=as.integer(targetLabels)
		
		if(!is.null(wi)) {
			if(is.null(names(wi)))
				stop("wi has to be a named vector!")
			
			if( !all(names(Wi)%in% names(wi)) )
				stop("Mismatch between provided names for 'wi' and class labels.")
			
			for(i in 1:length(wi)){
				Wi[names(wi)[i]]=wi[i]
			}
		}
	}
	
	# Cross-validation?
	if(cross<0){
		stop("Cross-validation argument 'cross' cannot be negative!")
	}
	else if(cross>n){
		stop("Cross-validation argument 'cross' cannot be larger than the number of samples (",n,").",sep="")
	}
	
	# Return storage preparation
	labels_ret = rep(0L, nbClass)
	nr_W = if(isMulticlass) nbClass else 1
	nc_W = if(bias) p+1 else p
	W_ret=matrix(data=0, nrow=nr_W, ncol=nc_W)
	
	# ACF check
	if (pref_min <= 0) {
        stop ("pref_min must be positive!")
    }
	if (pref_max <= 0) {
        stop ("pref_max must be positive!")
    }
    if (pref_min >= pref_max) {
        stop ("pref_max must be greater than pref_min!")
    }
	if (change_rate <= 0) {
        stop ("Change rate must be positive!")
    }
	if (max_iter <= 0) {
        stop ("Maximum iteration must be positive!")
    }
    
	
	#
	# </Arg preparation>
	# as.double(t(X)) corresponds to rewrite X as a nxp-long vector instead of a n-rows and p-cols matrix. Rows of X are appended one at a time.
	ret <- .C("trainLinear",
			as.double(W_ret),
			as.integer(labels_ret),
			as.double(if(sparse) data@ra else t(data)),
			as.double(yC),
			as.integer(n),
			as.integer(p),
			# sparse index info
			as.integer(sparse),
			as.integer(if(sparse) data@ia else 0),
			as.integer(if(sparse) data@ja else 0),

			as.double(b),
			as.integer(type),
			as.double(cost),
			as.double(epsilon),
			as.double(svr_eps),
			as.integer(nrWi),
			as.double(Wi),
			as.integer(WiLabels),
			as.integer(cross),
	
            as.integer(max_iter),
			
			# acf stuff
			as.double(change_rate),
			as.double(pref_min),
			as.double(pref_max),
			
			as.integer(verbose),
			PACKAGE="LiblineaR.ACF"
			)
	
	if(cross==0){
		labels_ret = if(isRegression) NULL else ret[[2]]
		
		# classNames is a lookup table for conversion between outputs of C code and initial levels of target
		classNames = {
			if(isRegression) NULL 
			else if(is.logical(targetLabels)) as.logical(labels_ret)
			else if(is.null(levels(targetLabels))) labels_ret 
			else factor(levels(targetLabels)[labels_ret], levels=levels(targetLabels))
		}
		
		W_ret = matrix(data=ret[[1]], nrow=nr_W, ncol=nc_W)
		colnames(W_ret) = c(
			if(!is.null(colnames(data))) colnames(data) else paste0("W",1:p),
			if(bias) "Bias" else c()
		)
		
		if(isMulticlass)
			rownames(W_ret) = classNames
		
		m=list()
		class(m)="LiblineaR.ACF"
		m$TypeDetail=types[type+1]
		m$Type=type
		m$W=W_ret
		m$Bias=bias
		m$ClassNames=classNames
		m$NbClass=nbClass
		return(m)
	}
	else{
		return(ret[[1]][1])
	}
}
